### ISSUE TITLE
create action to save the uploaded dataTable to the store

### ISSUE THREAD

**Issue by surchs** [neurobagel/annotation_tool#256]
When I upload a dataTable via the index/home page, the content of this file should be parsed and then fed to an action that populates/updates the appropriate state variables in the store. 

A couple of things should happen in response to uploading a new dataTable:

1. the Table should be parsed into a JS object
2. the table object should be stored in the store.state
3. the columnToCategory map should be updated to have the column names from the current dataTable
4. the data dictionary will be initialized to have keys for every column in the data table - that is: anything already there will be wiped and re-initialized

There are some additional mutations that are needed for this action:

- [x] a mutation to store a new dataTable in the state
- [x] a mutation to create a new data dictionary from the dataTable

**Comment by surchs** [comment]
Some other notes:

- in cypress I can upload files now: https://www.cypress.io/blog/2022/01/19/uploading-files-with-selectfile/. This should allow me to test the mutation being fired from the correct page

**Comment by jarmoza** [comment]
@surchs I am attaching my current version of `index.vue` for two reasons. The first is that regarding testing the index page, we must ensure that the names of the methods being called by the file upload button are unique from their actions counterparts – either that or the component attribute can call the action directly.

Secondly, my new mutation in #291 will need to be called in this action along with the call to the `getColumnNames` getter to passed along to that mutation – as we had discussed today in our huddle over the issue.

```
<template>

    <b-container fluid>

        <!-- Data table file loading area -->
        <b-row>
            <h2>{{ uiText.dataTableHeader }}</h2>
        </b-row>

        <!-- Shows file contents -->
        <b-row>
            <textarea
                :cols="textArea.height"
                :rows="textArea.width"
                v-model="stringifiedDataTable" />
        </b-row>

        <!-- Selects data table file (i.e. participants.tsv) -->
        <b-row>
            <file-selector
                data-cy="data-table-selector"
                :content-type="contentTypes.dataTable"
                @file-selected="setDataTable($event)" />
        </b-row>


        <!-- Data dictionary file loading area -->
        <b-row>
            <h2>{{ uiText.dataDictionaryHeader }}</h2>
        </b-row>

        <!-- Shows file contents -->
        <b-row>
            <textarea
                :cols="textArea.height"
                :rows="textArea.width"
                v-model="stringifiedDataDictionary" />
        </b-row>

        <b-row>
            <!-- Selects data dictionary file (i.e. participants.json) -->
            <file-selector
                data-cy="data-dictionary-selector"
                :content-type="contentTypes.dataDictionary"
                @file-selected="setDataDictionary($event)" />
        </b-row>

    </b-container>

</template>

<script>

    // Allows for calls to store actions
    import { mapActions } from "vuex";

    // Allows for reference to store data by creating simple, implicit getters
    import { mapGetters } from "vuex";

    // Allows for direct mutations of store data
    import { mapMutations } from "vuex";

    // Allows for reference to store data by creating simple, implicit getters
    import { mapState } from "vuex";

    export default {

        name: "IndexPage",

        data() {

            return {

                // Content types that are expected for the file selectors
                contentTypes: {

                    dataDictionary: "application/json",
                    dataTable: "text/tab-separated-values"
                },

                // Size of the file display textboxes
                textArea: {

                    height: 200,
                    width: 5
                },

                // Text for UI elements
                uiText: {

                    dataTableHeader: "Data table",
                    dataDictionaryHeader: "Data dictionary"
                }
            };
        },

        computed: {

            ...mapGetters([

                "getColumnNames"
            ]),

            ...mapState([

                "dataDictionary",
                "dataTable",
                "pageData"
            ]),

            stringifiedDataDictionary() {

                // 0. Return a blank string if there is no loaded data dictionary file
                if ( !this.$store.getters.isDataDictionaryLoaded ) {

                    return "";
                }

                // 1. Return a string version of the data dictionary file
                // NOTE: Defaults to json for now
                return JSON.stringify(this.dataDictionary.original, null, 4);
            },

            stringifiedDataTable() {

                // 0. Return a blank string is there is no loaded data table
                if ( !this.$store.getters.isDataTableLoaded ) {

                    return "";
                }

                // 1. Convert the tsv file data into a list of strings
                // NOTE: Defaults to tsv for now
                const textAreaArray = [Object.keys(this.dataTable.original[0]).join("\t")];
                for ( let index = 0; index < Object.keys(this.dataTable.original[0]).length; index++ ) {

                    textAreaArray.push(Object.values(this.dataTable.original[index]).join("\t"));
                }

                // 2. Return the tsv file data joined as one string
                return textAreaArray.join("\n");
            }
        },

        methods: {

            ...mapActions([

                "saveDataDictionary",
                "saveDataTable"
            ]),

            ...mapMutations([

                "createColumnToCategoryMap"
            ]),

            setDataDictionary(p_fileData) {

                // Update the store with json file data
                // NOTE: Defaults to json for now
                this.saveDataDictionary({

                    data: p_fileData.data,
                    filename: p_fileData.filename,
                    fileType: "json"
                });
            },

            setDataTable(p_fileData) {

                // 1. Update the store with tsv file data
                // NOTE: Defaults to tsv for now
                this.saveDataTable({

                    data: ( 0 === p_fileData.data.length ) ? null : p_fileData.data,
                    filename: p_fileData.filename,
                    fileType: "tsv"
                });

                // 2. Create a new map for linking table columns to annotation categories
                this.initializeColumnToCategoryMap(this.getColumnNames);
            }
        }
    };

</script>
```

**Comment by surchs** [comment]
@jarmoza and @rmanaem: I'm realizing that we didn't really discuss what's supposed to happen when you add a new dataTable. I can think of at least a couple of cases that might have to behave differently:

1. I just launched the app with nothing in it. Once I upload my table, I expect it to populate the state without regard for anything already there (because there isn't any)
2. I have (for some reason) already uploaded a data dictionary, and now I upload my dataTable. It just so happens that my dictionary doesn't have information for all of the columns. What should I do
3. I noticed mid-annotation that I loaded the wrong data table ....

I think it'd make sense to have a quick chat about this stuff to make sure we don't go overboard with complexity but also have it be functional.
